File: problems/27-segment-tree/001-range-sum-query-mutable/python/solution.py
class NumArray:
    def __init__(self, nums: list[int]):
        self.n = len(nums)
        self.tree = [0] * (4 * max(1, self.n))
        if self.n:
            self._build(nums, 1, 0, self.n - 1)

    def _build(self, nums, node, left, right):
        if left == right:
            self.tree[node] = nums[left]
            return
        mid = (left + right) // 2
        self._build(nums, node * 2, left, mid)
        self._build(nums, node * 2 + 1, mid + 1, right)
        self.tree[node] = self.tree[node * 2] + self.tree[node * 2 + 1]

    def update(self, index: int, val: int) -> None:
        if self.n:
            self._update(1, 0, self.n - 1, index, val)

    def _update(self, node, left, right, index, val):
        if left == right:
            self.tree[node] = val
            return
        mid = (left + right) // 2
        if index <= mid:
            self._update(node * 2, left, mid, index, val)
        else:
            self._update(node * 2 + 1, mid + 1, right, index, val)
        self.tree[node] = self.tree[node * 2] + self.tree[node * 2 + 1]

    def sumRange(self, left: int, right: int) -> int:
        if not self.n:
            return 0
        return self._query(1, 0, self.n - 1, left, right)

    def _query(self, node, left, right, query_left, query_right):
        if query_left <= left and right <= query_right:
            return self.tree[node]
        mid = (left + right) // 2
        total = 0
        if query_left <= mid:
            total += self._query(node * 2, left, mid, query_left, query_right)
        if query_right > mid:
            total += self._query(node * 2 + 1, mid + 1, right, query_left, query_right)
        return total

File: problems/27-segment-tree/001-range-sum-query-mutable/java/Solution.java
public class Solution {
    private int[] tree;
    private int n;

    public Solution(int[] nums) {
        n = nums.length;
        tree = new int[4 * Math.max(1, n)];
        if (n > 0) build(nums, 1, 0, n - 1);
    }

    private void build(int[] nums, int node, int left, int right) {
        if (left == right) {
            tree[node] = nums[left];
            return;
        }
        int mid = left + (right - left) / 2;
        build(nums, node * 2, left, mid);
        build(nums, node * 2 + 1, mid + 1, right);
        tree[node] = tree[node * 2] + tree[node * 2 + 1];
    }

    public void update(int index, int value) {
        if (n > 0) update(1, 0, n - 1, index, value);
    }

    private void update(int node, int left, int right, int index, int value) {
        if (left == right) {
            tree[node] = value;
            return;
        }
        int mid = left + (right - left) / 2;
        if (index <= mid) update(node * 2, left, mid, index, value);
        else update(node * 2 + 1, mid + 1, right, index, value);
        tree[node] = tree[node * 2] + tree[node * 2 + 1];
    }

    public int sumRange(int left, int right) {
        if (n == 0) return 0;
        return query(1, 0, n - 1, left, right);
    }

    private int query(int node, int left, int right, int queryLeft, int queryRight) {
        if (queryLeft <= left && right <= queryRight) return tree[node];
        int mid = left + (right - left) / 2;
        int sum = 0;
        if (queryLeft <= mid) sum += query(node * 2, left, mid, queryLeft, queryRight);
        if (queryRight > mid) sum += query(node * 2 + 1, mid + 1, right, queryLeft, queryRight);
        return sum;
    }
}
